scico/functional/_tvnorm.py

```diff
@@ -0,0 +1,138 @@
+
+    has_eval = True
+    has_prox = True
+
+    def __init__(self, ndims: Optional[int] = None):
+        r"""
+        Args:
+            ndims: Number of (trailing) dimensions of the input over
+                which to apply the finite difference operator. If
+                ``None``, differences are evaluated along all axes.
+        """
+        self.ndims = ndims
+        self.h0 = snp.array([1.0, 1.0]) / snp.sqrt(2.0)   # lowpass filter
+        self.h1 = snp.array([1.0, -1.0]) / snp.sqrt(2.0)  # highpass filter
+        self.l1norm = L1Norm()
+        self.G = None
+        self.W = None
+
+    def __call__(self, x: Array) -> float:
+        r"""Compute the anisotropic TV norm of an array."""
+        if self.G is None or self.G.shape[1] != x.shape:
+            if self.ndims is None:
+                ndims = x.ndim
+            else:
+                ndims = self.ndims
+            axes = tuple(range(ndims))
+            self.G = FiniteDifference(x.shape, input_dtype=x.dtype, axes=axes, circular=True)
+        return snp.sum(snp.abs(self.G @ x))
+
+    @staticmethod
+    def _shape(idx: int, ndims: int) -> Tuple:
+        """Construct a shape tuple.
+
+        Construct a tuple of size `ndims` with all unit entries except
+        for index `idx`, which has a -1 entry.
+        """
+        return (1,) * idx + (-1,) + (1,) * (ndims - idx - 1)
+
+    def prox(self, v: Array, lam: float = 1.0, **kwargs) -> Array:
+        r"""Approximate proximal operator of the anisotripic  TV norm.
+
+        Approximation of the proximal operator of the anisotropic TV norm,
+        computed via the method described in :cite:`kamilov-2016-parallel`.
+
+        Args:
+            v: Input array :math:`\mb{v}`.
+            lam: Proximal parameter :math:`\lam`.
+            kwargs: Additional arguments that may be used by derived
+                classes.
+        """
+        if self.ndims is None:
+            ndims = v.ndim
+        else:
+            ndims = self.ndims
+        K = 2 * ndims
+
+        if self.W is None or self.W.shape[1] != v.shape:
+            C0 = VerticalStack(  # Stack of lowpass filter operators for each axis
+                [
+                    CircularConvolve(
+                        self.h0.reshape(AnisotropicTVNorm._shape(k, ndims)),
+                        v.shape,
+                        ndims=self.ndims,
+                    )
+                    for k in range(ndims)
+                ]
+            )
+            C1 = VerticalStack(  # Stack of highpass filter operators for each axis
+                [
+                    CircularConvolve(
+                        self.h1.reshape(AnisotropicTVNorm._shape(k, ndims)),
+                        v.shape,
+                        ndims=self.ndims,
+                    )
+                    for k in range(ndims)
+                ]
+            )
+            # single-level shift-invariant Haar transform
+            self.W = VerticalStack((C0, C1), jit=True)
```

Comment:
We don't need it here because the shift-invariant version of the highest-resolution Haar decomposition is computed by convolution, without the downsampling that would follow for the standard (shift-variant) Haar decomposition.